platform/qemu/virt/platform.c

```diff
@@ -122,8 +127,29 @@ static int virt_timer_init(bool cold_boot)
 
 static int virt_system_down(u32 type)
 {
-	/* For now nothing to do. */
-	return 0;
+	struct sbi_scratch *local_scratch = NULL;
+	int ret = 0;
+
+	/* Tell the "finisher" that the simulation
+	 * was successful so that QEMU exits
+	 */
+	writew(VIRT_TEST_FINISHER_PASS, (void *)VIRT_TEST_ADDR);
+
+	/* The above didn't work, send an IPI on
+	 * everyone to hang
+	 */
+	local_scratch = sbi_scratch_thishart_ptr();
+
+	ret = sbi_ipi_send_many(local_scratch,
+				NULL, SBI_IPI_EVENT_HALT);
+
```

Comment:
It's good to send halt event in-case platform specific shutdown failed.

I think the sbi_ipi_send_many() call should called from generic library so that we have it for all platforms. You can add this to sbi_system_reboot() and sbi_system_shutdown() functions of lib/sbi_shutdown.c. You can use "lib: " prefix for changes in generic library.

Reply:
OK I'll make a new PR